### PR TITLE
Expose console and API externally

### LINKED ISSUE #66: prod cluster API server needs to be externally accessible
Currently users would need to be on VPN in order to interact with the API server. For production we'll need to be able to access the API server from outside VPN using the `api.shift.nerc.mghpcc.org` address.

### LINKED ISSUE #67: prod cluster OpenShift console needs to be externally accessible
Users need to be able to access the OpenShift console externally. If we apply the external route label to the existing openshift-console route that would give us:

https://console-openshift-console.apps.shift.nerc.mghpcc.org

We might want to consider adding a custom route to the console to shorten this to just `https://shift.nerc.mghpcc.org`

### PR DESCRIPTION
**Pull request by larsks**
This commit includes several changes to make the OpenShift console and API
available externally (without VPN access). Broadly, there are

- Modify the cluster-ingress configuration to use externally visible
  hostnames for the console, downloads, and oauth services
- Add explicit external routes (i.e, routes with a label that matches
  our external ingress service) for these services
- Add a proxy to the OpenShift API, exposed with an external route

With these changes in place, it is possible to access the console from
outside the VPN. Some URLs (e.g., when using the "copy login command" menu
option) will still use the internal, vpn-only hostnames.

Closes ocp-on-nerc/operations#67
Closes ocp-on-nerc/operations#66

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204165599358399

### CODE REVIEW

File: cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        - containerPort: 8080
+          name: http
+        volumeMounts:
+          - name: haproxy-config
+            mountPath: /usr/local/etc/haproxy
+          - name: haproxy-state
+            mountPath: /var/lib/haproxy
+        securityContext:
+          allowPrivilegeEscalation: false
+          runAsUser: 0
```
Comment (knikolla) [inline cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/deployment.yaml:23]: Why does this need to run as root?

File: cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/role.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
```
Comment (naved001) [inline cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/role.yaml:2]: Didn't know what this role was doing, and after some looking around I landed here: 
https://docs.openshift.com/container-platform/4.10/authentication/managing-security-context-constraints.html#role-based-access-to-ssc_configuring-internal-oauth

Question: Can non-admins create a role like this in their namespace and run their containers as root?

File: cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/routes/api.yaml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+apiVersion: route.openshift.io/v1
+kind: Route
+metadata:
+  name: api-external
```
Comment (naved001) [inline cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/routes/api.yaml:4]: I am confused, there's a route (cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/route.yaml) with the same name but to the service `api-proxy`. This one is to `apiserver` Service. Both have the same host though `api.apps.shift.nerc.mghpcc.org`.

Additionally, this is not included in the kustomization file in this directory.

File: cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/deployment.yaml (hunk 1 above)
Comment (larsks) [reply cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/deployment.yaml:23]: When I wrote this, I assumed that haproxy's default behavior of start as root and then suid-ing to an unprivileged user was intrinsic to haproxy, but it turns out that removing the `user` and `group` directories is sufficient.

File: cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/role.yaml (hunk 2 above)
Comment (larsks) [reply cluster-scope/overlays/nerc-ocp-prod/feature/external-console-access/api-proxy/role.yaml:2]: @naved001 if someone tries to create a rolebinding that grants permissions they don't already have, it will fail with:

```
Error from server (Forbidden): error when creating ".": roles.rbac.authorization.k8s.io "allow-anyuid" is forbidden: user "testuser" (groups=["system:authenticated:oauth" "system:authenticated"]) is attempting to grant RBAC permissions not currently held:
{APIGroups:["security.openshift.io"], Resources:["securitycontextconstraints"], ResourceNames:["anyuid"], Verbs:["use"]}
Error from server (NotFound): error when creating ".": roles.rbac.authorization.k8s.io "allow-anyuid" not found
```

**Comment by joachimweyl** [comment]
This was an example of one way to do this, we found a better way
